src/LanguageServer/Protocol/Handler/Diagnostics/BuildOnlyDiagnosticIdsHandler.cs

```diff
@@ -24,56 +23,33 @@ internal sealed record class BuildOnlyDiagnosticIdsResult([property: JsonPropert
 [method: ImportingConstructor]
 [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
 internal sealed class BuildOnlyDiagnosticIdsHandler(
-    DiagnosticAnalyzerInfoCache.SharedGlobalCache globalCache,
     [ImportMany] IEnumerable<Lazy<ILspBuildOnlyDiagnostics, ILspBuildOnlyDiagnosticsMetadata>> compilerBuildOnlyDiagnosticsProviders)
-                : ILspServiceRequestHandler<BuildOnlyDiagnosticIdsResult>
+    : ILspServiceRequestHandler<BuildOnlyDiagnosticIdsResult>
 {
     public const string BuildOnlyDiagnosticIdsMethodName = "workspace/buildOnlyDiagnosticIds";
 
-    private readonly DiagnosticAnalyzerInfoCache.SharedGlobalCache _globalCache = globalCache;
     private readonly ImmutableDictionary<string, string[]> _compilerBuildOnlyDiagnosticIds = compilerBuildOnlyDiagnosticsProviders
         .ToImmutableDictionary(lazy => lazy.Metadata.LanguageName, lazy => lazy.Metadata.BuildOnlyDiagnostics);
 
     public bool MutatesSolutionState => false;
     public bool RequiresLSPSolution => true;
 
-    public Task<BuildOnlyDiagnosticIdsResult> HandleRequestAsync(RequestContext context, CancellationToken cancellationToken)
+    public async Task<BuildOnlyDiagnosticIdsResult> HandleRequestAsync(RequestContext context, CancellationToken cancellationToken)
     {
-        Contract.ThrowIfNull(context.Solution);
+        var solution = context.Solution;
+        Contract.ThrowIfNull(solution);
 
-        using var _1 = ArrayBuilder<string>.GetInstance(out var builder);
-        foreach (var languageName in context.Solution.Projects.Select(p => p.Language).Distinct())
+        using var _ = ArrayBuilder<string>.GetInstance(out var builder);
+        foreach (var languageName in solution.Projects.Select(p => p.Language).Distinct())
         {
             if (_compilerBuildOnlyDiagnosticIds.TryGetValue(languageName, out var compilerBuildOnlyDiagnosticIds))
-            {
                 builder.AddRange(compilerBuildOnlyDiagnosticIds);
-            }
         }
 
-        using var _2 = PooledHashSet<(object Reference, string Language)>.GetInstance(out var seenAnalyzerReferencesByLanguage);
+        var diagnosticService = solution.Services.GetRequiredService<IDiagnosticAnalyzerService>();
+        builder.AddRange(await diagnosticService.GetCompilationEndDiagnosticDescriptorIdsAsync(
+            solution, cancellationToken).ConfigureAwait(false));
 
-        foreach (var project in context.Solution.Projects)
```

Comment:
instead of LSP handler calling into internal type and forcing loads locally, this now calls through diag service, which will apropriately oop if necessary.